### PR TITLE
Add bugs api

### PR DESCRIPTION
**Pull request by OliverHGray**

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                    items:
+                      type: object
+                      properties:
+                        bugId:
+                          type: string
+                        isFound:
+                          type: boolean
+                        foundDate:
+                          type: string
+                          description: ISO date string
```
Comment (Taklop) [inline swagger.yaml:33]: I do love me some swagger

File: src/app/index.ts [hunk 2]
```diff
@@ -3,14 +3,16 @@ import * as helmet from 'helmet';
 import * as cors from 'cors';
 import * as cookieParser from 'cookie-parser';
 import { auth } from 'firebase-admin';
+import { Repositories } from '../repository';
 import authenticate from './middleware/authenticate';
-import { Firestore } from '@google-cloud/firestore';
-import apis from '../apis';
```
Comment (Taklop) [inline src/app/index.ts:8]: I prefer the name apis to service in this case, but it's still okay

File: src/app/middleware/handleErrors.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import { ValidationError } from '@hapi/joi';
+
+export default (error: Error, _request: Request, response: Response, _next: NextFunction) => {
+    switch (error.name) {
+        case 'ValidationError':
+            return handleValidationError(error as ValidationError, response);
+        default:
+            return response.status(500).send('An unexpected error has occured.');
+    }
+}
```
Comment (Taklop) [inline src/app/middleware/handleErrors.ts:11]: Nice ol' catch all handler 👌

File: src/boot.ts [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import * as firebase from 'firebase-admin';
 import { Firestore } from '@google-cloud/firestore';
 import app from './app';
+import repository from './repository';
 
 const oauth = firebase.initializeApp().auth();
-const database = new Firestore();
+const repositories = repository(new Firestore());
```
Comment (Taklop) [inline src/boot.ts:7]: prefer db to repository but this is cool too

File: src/repository/bugs.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import { Firestore } from "@google-cloud/firestore";
+
+const COLLECTION_NAME = 'bugs';
+
+export default (database: Firestore): BugRepository => {
+    const collection = database.collection(COLLECTION_NAME);
```
Comment (Taklop) [inline src/repository/bugs.ts:6]: If this is only ever going to be the bugs collection, do we need the string extracted to a const?

It makes the line where we assign collection a bit less readable, as I have to read what COLLECTION_NAME actually is?

I'm probably missing something here

File: src/service/bugs/index.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+export default ({ bugs: repository }: { bugs: BugRepository }) => Router()
+    .get('', tsMiddleware()
+        .run(getBugs(repository))
+        .return(returnBugs)
+    )
+    .put('/:bugId', tsMiddleware()
+        .run(validate)
+        .run(getBugs(repository))
+        .return(updateBugState(repository))
```
Comment (Taklop) [inline src/service/bugs/index.ts:17]: Beautiful abstraction

File: src/service/bugs/updateBugState.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import * as dayjs from 'dayjs';
```
Comment (Taklop) [inline src/service/bugs/updateBugState.ts:1]: I heard mutability is more fun ;)
(Nice usage here over moment)

File: src/service/bugs/updateBugState.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    })
+    return toResponse(result);
+}
+
+const updateBugs = (bugs: Bug[], bugId: string, isFound: boolean) => {
+    const bugsObject = bugs.reduce((bugs: any, currentValue: Bug) => {
+        return {
+            ...bugs,
+            [currentValue.bugId]: {
+                isFound: true,
```
Comment (Taklop) [inline src/service/bugs/updateBugState.ts:20]: Ahhh, but you see Mr Oliver, what if I have mistakenly clicked a tick box, and need to untick it? We have stated that isFound is a boolean, but then assumed it to always be true. 😸

File: swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+          required: true
+      requestBody:
+        required: true
+        content:
+          application/json:
+            schema:
+              type: object
+              properties:
+                isFound:
+                  type: boolean
```
Comment (Taklop) [inline swagger.yaml:51]: Here we have said that the put takes a boolean, which implies we can "unfind" a bug. Which is true! I may restart the game, or mistakenly click a button etc.

File: test/integration/createApp.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                        return {
+                            exists: doc !== undefined,
+                            data: () => doc,
+                        }
+                    },
+                    update: (newDocument: any) => collection[documentPath] = newDocument,
+                }),
+            }
+        }
+    } as any;
```
Comment (Taklop) [inline test/integration/createApp.ts:58]: Can't believe google couldn't make a firestore emulator 😉

File: src/boot.ts (hunk 4 above)
Comment (OliverHGray) [reply src/boot.ts:7]: It used to be an instance of a database but now it's an instance of a repository

File: src/repository/bugs.ts (hunk 5 above)
Comment (OliverHGray) [reply src/repository/bugs.ts:6]: Not sure why I did this, I think to make it more clear if I copied and pasted the class to make a fish repository which parts I needed to change?

File: src/service/bugs/updateBugState.ts [hunk 11]
```diff
@@ -13,19 +13,24 @@ export default (repository: BugRepository) => async <TParams extends Params>({ b
...
-        }
-    }, {});
-    bugsObject[bugId] = {
-        isFound,
-        foundDate: dayjs().format('YYYY-MM-DD'),
+    const convertToBugsObject = () => bugs.reduce((accumulator: any, currentValue: Bug) => {
+            return {
+                ...accumulator,
+                [currentValue.bugId]: {
+                    isFound: true,
```
Comment (Taklop) [inline src/service/bugs/updateBugState.ts:20]: How do we set isFound to false @OliverHGray ?

**Pull request merged**